core/common/src/main/java/alluxio/security/authentication/PlainSaslServerCallbackHandler.java

```diff
@@ -67,8 +69,17 @@ public void handle(Callback[] callbacks) throws IOException, UnsupportedCallback
     mAuthenticationProvider.authenticate(username, password);
 
     if (ac != null) {
+      // TODO(gpang): check for impersonation permissions
+
       ac.setAuthorized(true);
 
+      try {
+        // getAuthorizedID() only works after the AuthorizeCallback is authorized
+        mImpersonationAuthenticator.authenticate(username, ac.getAuthorizedID());
+      } catch (Exception e) {
+        ac.setAuthorized(false);
```

Comment:
Is it normal to use exceptions for control flow with authentication?

Reply:
I don't know if it is normal, but our `AuthenticationProvider` also throws an `AuthenticationException` when authentication fails.

Also, the nice thing about this is that I can have different exception messages for different reasons for denial.